ocaml/xapi/helpers.ml

```diff
@@ -418,18 +423,22 @@ let make_rpc ~__context rpc : Rpc.response =
   in
   let http = xmlrpc ~subtask_of ~version:"1.1" path in
   let http = TraceHelper.inject_span_into_req tracing http in
-  let transport =
-    if Pool_role.is_master () then
-      Unix Xapi_globs.unix_domain_socket
-    else
-      SSL
-        ( SSL.make ~use_stunnel_cache:true ~verify_cert:(Stunnel_client.pool ())
-            ()
-        , Pool_role.get_master_address ()
-        , !Constants.https_port
-        )
-  in
-  dorpc ~srcstr:"xapi" ~dststr:"xapi" ~transport ~http rpc
+  match !rpc_fun with
+  | Some rpcfun when Pool_role.is_master () ->
+      rpcfun http rpc
+  | _ ->
+      let transport =
+        if Pool_role.is_master () then
+          Unix Xapi_globs.unix_domain_socket
+        else
+          SSL
+            ( SSL.make ~use_stunnel_cache:true
+                ~verify_cert:(Stunnel_client.pool ()) ()
+            , Pool_role.get_master_address ()
+            , !Constants.https_port
+            )
+      in
+      dorpc ~srcstr:"xapi" ~dststr:"xapi" ~transport ~http rpc
```

Comment:
The code could avoid reading the role from FS twice:
```suggestion
  match !rpc_fun, Pool_role.is_master () with
  | Some rpcfun, true ->
      rpcfun http rpc
  | None, true ->
      let transport = Unix Xapi_globs.unix_domain_socket in
      dorpc ~srcstr:"xapi" ~dststr:"xapi" ~transport ~http rpc
  | _, false ->
      let transport =
        SSL
            ( SSL.make ~use_stunnel_cache:true
                ~verify_cert:(Stunnel_client.pool ()) ()
            , Pool_role.get_master_address ()
            , !Constants.https_port
            )
      in
      dorpc ~srcstr:"xapi" ~dststr:"xapi" ~transport ~http rpc
```

Reply:
Reading the pool role should be cached, it is only read from the FS the first time. Although looks like I could optimize that function for the fastpath when the role is already set.

Reply:
https://github.com/xapi-project/xen-api/pull/6279 PR here to optimize `Pool_role.is_master`, it now takes ~7ns to call it.